### PR TITLE
homework 1.2 - implement login and get product calls

### PR DESCRIPTION
**Pull request by mtsoneva**

### CODE REVIEW

File: app/src/main/java/com/example/androidtraining/networking/API.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import com.example.androidtraining.models.Product
+import com.example.androidtraining.models.UserInfo
+import retrofit2.http.Body
+import retrofit2.http.GET
+import retrofit2.http.Headers
+import retrofit2.http.POST
+
+interface API {
+    @POST(value = "/api/auth/local")
+    @Headers("No-Authentication: true")
```
Comment (yll090) [inline app/src/main/java/com/example/androidtraining/networking/API.kt:13]: Is the header really necessary? It should work without it, too.

File: app/src/main/java/com/example/androidtraining/MainActivity.kt [hunk 2]
```diff
@@ -13,11 +14,17 @@ import androidx.compose.ui.Modifier
...
 import com.example.androidtraining.screens.login.LoginScreen
 import com.example.androidtraining.screens.product.ProductScreen
 import com.example.androidtraining.ui.theme.AndroidTrainingTheme
+import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
 
+@AndroidEntryPoint
 class MainActivity : ComponentActivity() {
+    @Inject
+    lateinit var tokenProvider: TokenProvider
```
Comment (yll090) [inline app/src/main/java/com/example/androidtraining/MainActivity.kt:27]: `TokenProvider` should be injected in a viewModel directly. The idea is that the view doesn't need to know about anything non-view/(business) logic-related stuff. You can safely do that and figure out all the logic in the viewModel. The login should only be notified if there is an error or if there is a successful login.

File: app/src/main/java/com/example/androidtraining/networking/APIClient.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    // User and Auth service
+    private fun createService(): API {
+        val okHttpClient: OkHttpClient = OkHttpClient.Builder()
+            .addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
+            .connectTimeout(DEFAULT_TIMEOUT, TimeUnit.MINUTES)
+            .readTimeout(DEFAULT_TIMEOUT, TimeUnit.MINUTES)
+            .writeTimeout(DEFAULT_TIMEOUT, TimeUnit.MINUTES)
+            .addInterceptor { chain ->
```
Comment (yll090) [inline app/src/main/java/com/example/androidtraining/networking/APIClient.kt:25]: This is redundant.

File: app/src/main/java/com/example/androidtraining/screens/product/ProductScreen.kt [hunk 4]
```diff
@@ -15,56 +16,91 @@ import androidx.compose.material3.Button
...
+    val viewModel: MainViewModel = hiltViewModel()
+    val response by viewModel.productResponse.collectAsState()
+    val context = LocalContext.current
+    val productData = remember { mutableStateOf<Product?>(null) }
+    LaunchedEffect(true) {
+        viewModel.getProduct()
+    }
 
+    LaunchedEffect(response) {
+        response?.let { result ->
```
Comment (yll090) [inline app/src/main/java/com/example/androidtraining/screens/product/ProductScreen.kt:45]: Continuation of the previous comment -- most of this logic this should happen in the viewModel.

File: app/src/main/java/com/example/androidtraining/networking/TokenProvider.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.example.androidtraining.networking
+
+class TokenProvider(
+    private var jwtToken: String? = null,
+    private var isLoggedIn: Boolean = false
+) {
+
+    fun setJwtToken(token: String) {
+        jwtToken = token
```
Comment (yll090) [inline app/src/main/java/com/example/androidtraining/networking/TokenProvider.kt:9]: What if we want to clear the token? After all, the local variable is nullable. Additionally, the separate boolean is also not really needed -- either the user has a token, or s/he doesn't.

**Review by yll090 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/androidtraining/networking/API.kt (hunk 1 above)
Comment (mtsoneva) [reply app/src/main/java/com/example/androidtraining/networking/API.kt:13]: removed

File: app/src/main/java/com/example/androidtraining/networking/APIClient.kt (hunk 3 above)
Comment (mtsoneva) [reply app/src/main/java/com/example/androidtraining/networking/APIClient.kt:25]: removed

File: app/src/main/java/com/example/androidtraining/MainActivity.kt (hunk 2 above)
Comment (mtsoneva) [reply app/src/main/java/com/example/androidtraining/MainActivity.kt:27]: fixed

File: app/src/main/java/com/example/androidtraining/networking/TokenProvider.kt (hunk 5 above)
Comment (mtsoneva) [reply app/src/main/java/com/example/androidtraining/networking/TokenProvider.kt:9]: I have created fun clearToken() which I intend to use after logout. Or I can use setJwtToken with null value.
For additional bool I agree - will remove it.

**Comment by yll090** [comment]
@mtsoneva there are two more comments to be resolved. 
1. You should be using the viewModel data directly instead of assigning it to a local state variable in the view (productData) -- that is why the viewModel is all about.
2. Make sure to not rely on the responses in the view (the jwt token change is correct) -- the view shouldn't know the logic behind the user being successfully logged in or not.

**Comment by mtsoneva** [comment]
@yll090 
1 is clear for me
what do you mean by "not rely on the responses in the view"? Can I use the response in the view in this way:

    LaunchedEffect(response) {
        response?.let { result ->
            result.onFailure {
                Toast.makeText(context, "Error", Toast.LENGTH_SHORT).show()
            }
        }
    }

So I can show the toast on error?

**Comment by yll090** [comment]
My point is to decouple the login success/error from the response directly. The login method can setup a flag or an enum or callbacks to listen for changes to the state of the login. This way if you need later on to add more logic to the setup you can do it safely without relying on the login response directly.

**Review by yll090 — APPROVED** [review]

File: app/src/main/java/com/example/androidtraining/MainViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class MainViewModel @Inject constructor(private val firstRepository: FirstRepository) :
+    ViewModel() {
+    @Inject
+    lateinit var tokenProvider: TokenProvider
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/androidtraining/MainViewModel.kt:21]: Note: You already have an `@Inject` constructor. You can have the `tokenProvider` there also.

File: app/src/main/java/com/example/androidtraining/models/ProductModel.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package com.example.androidtraining.models
+
+data class Product(
+    val id: Number,
+    val title: String,
+    val description: String,
+    val short_description: String,
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/androidtraining/models/ProductModel.kt:7]: You should use `@SerializedName` here. `snake_case` is an invalid variable syntax in Kotlin.

File: app/src/main/java/com/example/androidtraining/MainViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import com.example.androidtraining.models.UserInfo
+import com.example.androidtraining.networking.TokenProvider
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class MainViewModel @Inject constructor(private val firstRepository: FirstRepository) :
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/androidtraining/MainViewModel.kt:18]: There should be no such thing as a `Main` viewmodel. Each screen should have its own dedicated viewmodel, positioned directly to it file-wise in the project folder structure.

File: app/src/main/java/com/example/androidtraining/MainViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    private val _response = MutableStateFlow<Result<LoginResponse>?>(null)
+    val response: StateFlow<Result<LoginResponse>?> = _response
+
+    private val _successfulLogin = MutableStateFlow<Boolean?>(null)
+    val successfulLogin: StateFlow<Boolean?> = _successfulLogin
+
+    private val _isLoading = MutableStateFlow(false)
+    var isLoading: StateFlow<Boolean> = _isLoading
+
+    fun login(userInfo: UserInfo) {
```
Comment (gbuyukliev-accedia) [inline app/src/main/java/com/example/androidtraining/MainViewModel.kt:31]: Note: Structure your code in accordance to types. A general guide is properties > custom constructors (if any) > functions. Having a function in the middle of property declarations is reducing code clarity.

**Review by gbuyukliev-accedia — COMMENTED** [review]
Just some post factum comments as I couldn't get to review it in time.

**Pull request merged**